app/controllers/contacts_controller.rb

```diff
@@ -0,0 +1,38 @@
+class ContactsController < ApplicationController
+  before_action do
+    raise "missing provider code" unless params[:provider_code]
+  end
```

Comment:
Do we need this as a before_action?

Could it not be a guard added to the `provider` private method?